lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs

```diff
@@ -537,42 +461,122 @@ getLegacyTransaction tid = do
+    -- INPUT-UTXO        = IDX VALUE FRAGMENT-ID
+    -- INPUT-ACCOUNT     = %xff VALUE UNTAG-ACCOUNT-ID
+    -- IDX               = %x00-fe
+    putInput (TxIn inputId inputIx, inputValue) = do
+        -- TODO invariant about inputIx /= 0xff
         putWord8 . toEnum . fromEnum $ inputIx
-        putWord64be $ getCoin coin
+        putWord64be $ getCoin inputValue
         putByteString $ getHash inputId
 
-    putOutput (TxOut address coin) = do
-        putAddress address
-        putWord64be $ getCoin coin
+    -- OUTPUT            = ADDRESS VALUE
+    putOutput (TxOut outputAddr outputValue) = do
+        putAddress outputAddr
+        putWord64be $ getCoin outputValue
+
+    -- WITNESS           = WITNESS-OLDUTXO / WITNESS-UTXO / WITNESS-ACCOUNT / WITNESS-MULTISIG
+    -- WITNESS-OLDUTXO   = %x00 LEGACY-XPUB LEGACY-SIGNATURE
+    -- WITNESS-UTXO      = %x01 ED25519-SIGNATURE
+    -- WITNESS-ACCOUNT   = %x02 SINGLE-ACNT-SIG
+    -- WITNESS-MULTISIG  = %x03 MULTI-ACNT-SIG
+    putWitness (xprv, Passphrase pwd) = \case
+        tag@TxWitnessUTxO -> do
+            putTxWitnessTag tag
+            putByteString $ unXSignature $ sign pwd xprv msg
+        tag@TxWitnessLegacyUTxO -> do
+            putTxWitnessTag tag
+            putByteString $ unXPub $ toXPub xprv
+            putByteString $ unXSignature $ sign pwd xprv msg
+        TxWitnessAccount ->
+            error "putWitness: TxWitnessAccount: not implemented"
+        TxWitnessMultisig ->
+            error "putWitness: TxWitnessMultisig: not implemented"
+      where
+        msg = mconcat
+            [ block0Hash
+            , blake2b256 $ BL.toStrict $ runPut putInputsOutputs
+            ]
+
+    -- DLG-CERT         = UNTAG-ACCOUNT-ID DLG-TYPE
+    -- DLG-TYPE         = DLG-NONE / DLG-FULL / DLG-RATIO
+    -- DLG-NONE         = %x00
+    -- DLG-FULL         = %x01 POOL-ID
+    -- DLG-RATIO        = %x02-FF %x02-08 2*8DLG-RATIO-POOL
+    -- DLG-RATIO-POOL   = %x01-FF POOL-ID
+    -- POOL-ID          = 32OCTET
+    putDelegationCertificate (ChimericAccount accountId) = \case
+        tag@DlgNone -> do
+            putByteString accountId
+            putWord8 (stakeDelegationTypeTag tag)
+        tag@(DlgFull (PoolId poolId)) -> do
+            putByteString accountId
+            putWord8 (stakeDelegationTypeTag tag)
+            putByteString poolId
+        DlgRatio ->
+            error "putDelegationCertificate: DlgCertRatio: not implemented"
+
+    -- STAKE-AUTH       = ACCOUNT-SIG
+    -- ACCOUNT-SIG      = %x01 SINGLE-ACNT-SIG
+    -- SINGLE-ACNT-SIG  = ED25519-SIGNATURE
+    putStakeAuthentication (ChimericAccount accountId) (xprv, Passphrase pwd) = do
+        putWord8 1
+        putByteString $ unXSignature $ sign pwd xprv accountId
+
+sealFragment :: Put -> (Hash "Tx", SealedTx)
+sealFragment put =
+    -- NOTE:
+    -- The fragment id is a hash of the fragment, minus the size (first 2 bytes)
+    (Hash (blake2b256 $ BS.drop 2 bytes), SealedTx bytes)
```

Comment:
😕

Reply:
Yeah. Not so happy either. That's provisional, hopefully. I am still working on getting the tests to run.